### ISSUE TITLE
Exclusively use systemd unit files to manage containers

### ISSUE THREAD

**Issue by smashwilson** [deconst/deconst-docs#86]
The docker containers that Ansible launches don't persist across system reboots. To start a container that lasts beyond the first CoreOS automatic update, we need to install and enable systemd unit files that launch and manage each one.

I'm doing both right now, but it's ugly and there's a risk of drift between the container configurations specified by each. I believe the way forward is to use systemd unit files exclusively, and possibly:
- Install them with cloud-config rather than Ansible
- Manage them with fleet

The questions that I'll need answers to:
- How do we add the public ports of each launched container to the load balancer? I'm computing a predictable exposed port for each now, but I'd prefer to let Docker decide the public port and register it dynamically.
- How can I orchestrate the deployment of new container image versions? Can it be done automatically on container image build? One option here is to go full-on immutable infrastructure, and "upgrade" by launching a new cluster and taking an axe to the old one once it's up. Another is to pull on service start, and trigger service restarts on update.
- We'll need to use a non-static discovery mechanism for etcd.

**Comment by rgbkrk** [comment]
I'm a little late in saying this, but you could use fleet files that are global units. That would also make any new node get that fleet service to run on itself.

**Comment by smashwilson** [comment]
> I'm a little late in saying this...

It's all just code :wink: As long as we have a no-downtime migration path, we can do whatever.

I'd prefer to keep the existing cluster stable for at least this week, but I'm already thinking about next steps. Any brainstorming help or insights you've got are appreciated.

> ... but you could use fleet files that are global units. That would also make any new node get that fleet service to run on itself.

I've been looking (belatedly) more closely at Fleet. Do you know how it works with linked containers? Or do you need to introduce something like weave or flannel as a networking layer?

**Comment by rgbkrk** [comment]
Fleet doesn't (by itself) under stand linking, you have to be very specific in your unit files.

For instance, I run a (named) nbcache and nbviewer fleet fleet file on every single one of our CoreOS nodes, linking nbviewer to the running nbcache (https://gist.github.com/rgbkrk/f27c22d0dc414357a1e3#file-nbviewer-service):

```
ExecStart=/usr/bin/docker run --name nbviewer \
      -e GITHUB_OAUTH_KEY=SOMEKEY \
      -e GITHUB_OAUTH_SECRET=SOMESECRET \
      --link nbcache:nbcache \
      -p 80:8080 \
      jupyter/nbviewer
```

The important part to get right here is the very beginning of nbviewer.service:

```
[Unit]
Description=NotebookViewer
BindsTo=nbcache.service

Requires=docker.service
Requires=etcd.service

After=docker.service
After=etcd.service
After=nbcache.service
```

This makes sure that the nbviewer service doesn't go up until nbcache.service is.

You could do all this with regular systemd unit files, but the magic is that with the fleet approach + global units any new node gets automatically configured. You can even have it do shenanigans with LBs, DNS, etc. That's how nbviewer runs now.

**Comment by smashwilson** [comment]
I actually read up on systemd a bunch to fix #80, so I did know about the `Requires` and `After` unit file directives; you can already see it in [the unit files for services with dependencies](https://github.com/deconst/deploy/blob/master/roles/worker/templates/presenter.service.j2#L5-L6). It'll be handy to have the nbviewer setup as a reference as we move forward, though :+1: I'll have to steal that dangling image cleanup, in particular :grin:

Here's my wish list for whatever deployment strategy we settle on:

**Anyone should be able to go from having nothing to a fully working system in three steps: clone the deploy repository; copy and edit a single file; run a single script.** The single file is `credentials.yml`, which has your Rackspace and Mongo credentials as well as a bunch of instance customization parameters that you can tune. We're close to this now, although you do need to have a reasonable Python install with Ansible and docker-py installed first.

**Bringing an existing cluster up to date should be done with zero downtime, and also done with a single script.** The Ansible stuff does this via `state=reloaded` on the Docker modules. We could do this with Fleet/systemd by just doing rolling service restarts, like you did with nbviewer.

**Scaling an existing cluster up and down should be possible by editing a single variable and running a single script.** This one we don't satisfy right now because newly added hosts won't join the etcd cluster properly; see #69.

As long as we satisfy these three properties, I personally don't mind ditching Ansible entirely and rolling something with Fleet and bash scripts, or using custom Python scripts or something like that. I expect we'll settle on some kind of hybrid because we're close to that anyway.

**Comment by smashwilson** [comment]
> You can even have it do shenanigans with LBs, DNS, etc.

How are you handling adding and removing nodes from the load balancer at the right times? I think that's the biggest piece that I'm missing from a full Fleet solution.

Even with Ansible I'm not particularly happy with my solution for this one: I'm computing [predictable external port numbers](https://github.com/deconst/deploy/blob/master/roles/worker/templates/content.service.j2#L21) for each service within each pod and [mapping those to the LB](https://github.com/deconst/deploy/blob/master/roles/loadbalanced/tasks/main.yml#L10). It's ugly, I'd much rather let Docker assign ports automatically and add those, but:
1. As ansible/ansible-modules-core#1428 says, the docker module doesn't currently pass ports as part of the registered output! Whoops.
2. I'm not sure how to remove LB nodes from hosts that are brought down after draining connections and so on.

You have some kind of Go service that handles this at the application layer, didn't you?

**Comment by rgbkrk** [comment]
> You have some kind of Go service that handles this at the application layer, didn't you?

Before switching to Fastly (which continually checks health), I used [peekaboo](https://github.com/rgbkrk/peekaboo). I had a fleet file for that too which is likely in an old commit of that nbviewer setup.

**Comment by smashwilson** [comment]
Shipped in deconst/deploy#46.
